src/main/scala-2.11/markets/clearing/engines/CDAMatchingEngineLike.scala

```diff
@@ -0,0 +1,104 @@
+
+  protected val bidOrdering: PriceOrdering[BidOrder]
+
+  protected var orderBook: immutable.Set[Order]
+
+  def askOrderBook: immutable.Seq[AskOrder] = {
+    val askOrders = orderBook.collect { case order : AskOrder => order }
+    askOrders.to[immutable.Seq].sorted(askOrdering)
+  }
+
+  def bidOrderBook: immutable.Seq[BidOrder] = {
+    val bidOrders = orderBook.collect { case order : BidOrder => order }
+    bidOrders.to[immutable.Seq].sorted(bidOrdering)
+  }
+
+  protected def crosses(incomingOrder: Order, existingOrder: Order): Boolean = {
+    (incomingOrder, existingOrder) match {
+      case (ask: AskOrder, bid: BidOrder) => ask.price <= bid.price
+      case (bid: BidOrder, ask: AskOrder) => bid.price >= ask.price
+    }
+  }
+
+  def fill(incomingOrder: Order): Option[immutable.Iterable[FilledOrder]] = {
+
+    /** Accumulate some filled orders
+      *
+      * @param incomingOrder
+      * @param filledOrders
+      * @param existingOrders CALL BY NAME!
+      * @return
+      */
+    @tailrec
+    def accumulate(incomingOrder: Order,
+                   filledOrders: immutable.Queue[FilledOrder],
+                   existingOrders: => immutable.Seq[Order]): immutable.Queue[FilledOrder] = {
+      existingOrders.headOption match {
+        case Some(existingOrder) if crosses(incomingOrder, existingOrder) =>
+          orderBook -= existingOrder  // SIDE EFFECT!
+          val residualQuantity = incomingOrder.quantity - existingOrder.quantity
+          val price = formPrice(incomingOrder, existingOrder)
+          val quantity = math.min(incomingOrder.quantity, existingOrder.quantity)
+          if (residualQuantity < 0) {
+            val filledOrder = TotalFilledOrder(incomingOrder.issuer, existingOrder.issuer, price, quantity, 1, incomingOrder.tradable)
+            // add residualOrder back into orderBook!
+            val residualOrder = existingOrder.split(-residualQuantity)
+            orderBook += residualOrder  // SIDE EFFECT!
+            filledOrders.enqueue(filledOrder)
+          } else if (residualQuantity == 0) {  // no rationing for incoming order!
+              val filledOrder = TotalFilledOrder(incomingOrder.issuer, existingOrder.issuer, price, quantity, 1, incomingOrder.tradable)
+              filledOrders.enqueue(filledOrder)
+          } else {  // incoming order is larger than existing order and will be rationed!
+            val filledOrder = PartialFilledOrder(incomingOrder.issuer, existingOrder.issuer, price,
+              quantity, 1, incomingOrder.tradable)
+            val residualOrder = incomingOrder.split(residualQuantity)
+            accumulate(residualOrder, filledOrders.enqueue(filledOrder), existingOrders)
+          }
+        case _ => // existingOrders is empty or incoming order does not cross best existing order.
+          orderBook += incomingOrder  // SIDE EFFECT!
+          filledOrders
+      }
+    }
+
+    incomingOrder match {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Default case not specified.](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1425046359/issues/source?bid=2921362&fileBranchId=2987952#l93)